frontends/p4/moveDeclarations.cpp

```diff
@@ -142,11 +159,48 @@ const IR::Node *MoveInitializers::postorder(IR::Declaration_Variable *decl) {
 }
 
 const IR::Node *MoveInitializers::postorder(IR::ParserState *state) {
-    if (oldStart == nullptr) return state;
-    if (state->name == IR::ParserState::start) state->name = newStartName;
+    if (state->name == IR::ParserState::start && !toMove->empty()) {
+        if (!loopsBackToStart) {
+            // Just prepend all initializing assignments to the original start
+            // state if there are no loops back to the start state.
+            state->components.insert(state->components.begin(), (*toMove).begin(), (*toMove).end());
+            toMove = new IR::IndexedVector<IR::StatOrDecl>();
+        } else {
+            // Otherwise, we need to insert all such assignments into a new start state
+            // that can only run once, so that the assignments are only executed one time.
+            state->name = newStartName;
+        }
+    }
+
     return state;
 }
 
+const IR::Node *MoveInitializers::postorder(IR::Path *path) {
+    if (!oldStart || !loopsBackToStart || path->name != IR::ParserState::start) return path;
+
+    // Only rename start state references if the parser contains initializing assignments
+    // and loops back to the start state.
+    auto decl = getDeclaration(getOriginal()->to<IR::Path>());
+    if (!decl->is<IR::ParserState>()) return path;
+    return new IR::Path(newStartName);
+}
+
+const IR::Node *MoveInitializers::postorder(IR::P4Parser *parser) {
+    if (oldStart && loopsBackToStart) {
+        // Only add a new state if the parser has initializers and contains one
+        // or more states that loop back to the start state.
+        auto newStart = new IR::ParserState(IR::ID(IR::ParserState::start), *toMove,
+                                            new IR::PathExpression(newStartName));
+        toMove = new IR::IndexedVector<IR::StatOrDecl>();
+        parser->states.insert(parser->states.begin(), newStart);
+    }
+
+    oldStart = nullptr;
+    loopsBackToStart = false;
```

Comment:
They are already initialized in the `MoveInitializers` constructor. Are you suggesting to not initialize them there at all, or to initialize them twice (in both constructor and preorder function) the first time?

Reply:
I think initialization in preorder makes sense, whether to initialize also in constructor is up to you. Some linters may complain about unitialized class fields if the constructor does not initialize it (because they simply have no way of proving that `preorder`/`init_apply`/etc. runs before anything else that uses the var), but I think having no initializer in constructor makes a little more sense as the value is then overwritten anyway.